### DISCUSSION TITLE
Error while deploying with Laravel from my GitLab repository. Issue with libssl.so.10 (PHP on Windows).

### DISCUSSION THREAD

**Discussion by DramaCity** [vercel/community#6279] [Help] [answered]
### Summary

I'm new to Vercel and have been trying to deploy my project from GitLab using the import feature. Initially, it was working, but when I clicked on the link, it downloaded a file instead of opening my site. I then created the api/index.php, vercel.json, and .vercelignore files, but now I'm encountering this error and I can't figure it out. Here is the build log and error message:

Running build in Washington, D.C., USA (East) – iad1
20:18:08.884 | Cloning gitlab.com/devjp1/tp1h24_web-sec_jasminm_pierrejrg (Branch: main, Commit: b702c1e)
20:18:09.230 | Previous build cache not available
20:18:09.904 | Cloning completed: 1.019s
20:18:10.437 | Found .vercelignore
20:18:10.439 | Removed 0 ignored files defined in .vercelignore
20:18:10.717 | Running "vercel build"
20:18:11.220 | Vercel CLI 33.6.1
20:18:11.847 | > Installing Builder: vercel-php@0.5.1
20:18:14.930 | Installing dependencies...
20:18:16.357 | npm WARN deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
20:18:28.611 |  
20:18:28.611 | added 73 packages in 13s
20:18:28.612 |  
20:18:28.612 | 9 packages are looking for funding
20:18:28.612 | run `npm fund` for details
20:18:28.630 | Detected `package-lock.json` generated by npm 7+
20:18:28.630 | Running "npm run build"
20:18:28.920 |  
20:18:28.922 | > build
20:18:28.922 | > vite build
20:18:28.923 |  
20:18:29.165 | vite v5.1.6 building for production...
20:18:29.202 | transforming...
20:18:29.783 | ✓ 47 modules transformed.
20:18:29.853 | rendering chunks...
20:18:29.866 | computing gzip size...
20:18:29.876 | public/vite/manifest.json              0.26 kB │ gzip:  0.15 kB
20:18:29.876 | public/vite/assets/main-B59N90_Z.css   4.59 kB │ gzip:  1.15 kB
20:18:29.878 | public/vite/assets/app-CifqVuM1.js    29.83 kB │ gzip: 12.03 kB
20:18:29.879 | ✓ built in 687ms
20:18:29.932 | 🐘 Downloading user files
20:18:29.932 | 🐘 Downloading PHP runtime files
20:18:29.938 | 🐘 Installing Composer dependencies [START]
20:18:29.945 | php: error while loading shared libraries: libssl.so.10: cannot open shared object file: No such file or directory
20:18:29.951 | Error: Exited with 127
20:18:29.951 | at ChildProcess.<anonymous> (/vercel/path0/.vercel/builders/node_modules/vercel-php/dist/utils.js:182:24)
20:18:29.952 | at ChildProcess.emit (node:events:518:28)
20:18:29.952 | at ChildProcess.emit (node:domain:488:12)
20:18:29.952 | at ChildProcess._handle.onexit (node:internal/child_process:294:12)
20:18:30.001 | Error: Command "vercel build" exited with 1
20:18:30.069 | Command "vercel build" exited with 1

### Steps to Reproduce

I created a laravel project, added mongodb ext for php on my system and the extension mongodb/laravel-mongodb. php v8.1.27 
api/index.php
<?php
/**
 * Here is the serverless function entry
 * for deployment with Vercel.
 */
require __DIR__.'/../public/index.php';

vercel.json
{
    "version": 2,
    "framework": null,
    "functions": {
        "api/index.php": { "runtime": "vercel-php@0.5.1", "includeFiles": "vendor/**" }
    },
    "routes": [{
        "src": "/(.*)",
        "dest": "/api/index.php"
    }],
    "env": {
        "APP_ENV": "production",
        "APP_DEBUG": "true",
        "APP_URL": "https://yourproductionurl.com",

        "APP_CONFIG_CACHE": "/tmp/config.php",
        "APP_EVENTS_CACHE": "/tmp/events.php",
        "APP_PACKAGES_CACHE": "/tmp/packages.php",
        "APP_ROUTES_CACHE": "/tmp/routes.php",
        "APP_SERVICES_CACHE": "/tmp/services.php",
        "VIEW_COMPILED_PATH": "/tmp",

        "CACHE_DRIVER": "array",
        "LOG_CHANNEL": "stderr",
        "SESSION_DRIVER": "cookie"
    },
    "build": {
        "env" : {
            "EXTENSIONS": "curl,mongodb,openssl"
        }
    }
}

**Comment by emre-koc** [comment]
Hi @rauchg Did vercel remove support for php ?

**Comment by amyegan** [comment]
Hey there! PHP is supported through a [community runtime](https://vercel.com/docs/functions/runtimes#community-runtimes) which is publicly available at [github.com/vercel-community/php](https://github.com/vercel-community/php). 

Please make sure the project is configured as recommended in the docs to ensure your PHP file is interpreted as a serverless function rather than a downloadable file.

[vercel.com/docs/functions/configuring-functions/runtime#other-runtimes](https://vercel.com/docs/functions/configuring-functions/runtime#other-runtimes).

**Reply by DramaCity** [reply]
Hi

I've completed all the necessary configuration for deployment, but the error persists. I'll edit the post to add the api/index.php and the vercel.json files in case the error is there, but I'm pretty sure it has to do with a dependency that I'm not aware of, which requires libssl. However, I'm pretty sure it has to do with the vercel-php environment which doesn't have libssl.

**Reply by emre-koc** [reply]
@amyegan this is the same case on our side, 
vercel-php is using @libphp/amazon-linux-2-v83 image and that image doesn't seem to have libssl.so.10. That is the reason for this error.

**Reply by emre-koc** [reply]
@amyegan I'm not sure if you fixed something but I was able to deploy laravel 10 finally.

**Reply by DramaCity** [reply]
did you need to add something to the deployment to make it work, Because it still doesn't work for me? (I only added the output directory to public, I didn't add any environment variables)

**Reply by emre-koc** [reply]
I'm using this vercel.json file. Hope this helps

```
{
    "version": 2,
    "framework": null,
    "functions": {
        "api/index.php": {
            "runtime": "vercel-php@0.7.0"
        }
    },
    "routes": [
        {
            "src": "/(favicon\\.ico|robots\\.txt)",
            "dest": "/public/$1"
        },
        {
            "src": "/build/(.*)",
            "dest": "/public/build/$1"
        },
        {
            "src": "/(.*)",
            "dest": "/api/index.php"
        }
    ],
    "env": {
        "APP_NAME": "api.treckstar.net",
        "APP_ENV": "production",
        "APP_DEBUG": "false",
        "APP_URL": "https://api.treckstar.net",
        "VERCEL_DEMO_MODE": "true",
        "APP_CONFIG_CACHE": "/tmp/config.php",
        "APP_EVENTS_CACHE": "/tmp/events.php",
        "APP_PACKAGES_CACHE": "/tmp/packages.php",
        "APP_ROUTES_CACHE": "/tmp/routes.php",
        "APP_SERVICES_CACHE": "/tmp/services.php",
        "CACHE_DRIVER": "array",
        "LOG_CHANNEL": "stderr",
        "SESSION_DRIVER": "array",
        "VIEW_COMPILED_PATH": "/tmp/views",
        "SSR_TEMP_PATH": "/tmp/ssr",
        "NODE_PATH": "node"
    }
}
```

**Reply by emre-koc** [reply]
and my api/index.php file is below. Reason I modify $_SERVER variables is Laravel couldn't get correct paths for routing. These updates fixed the issue. Since we are not modifying public/index.php you can run Laravel locally as usual. 

```
<?php

//IF $_SERVER['DOCUMENT_ROOT'] does not contain the public folder, add it
if (strpos($_SERVER['DOCUMENT_ROOT'], "/public") === false) {
    $_SERVER['SCRIPT_NAME'] = '/index.php';
    $_SERVER['DOCUMENT_ROOT'] = $_SERVER['DOCUMENT_ROOT']."/public";
    $_SERVER['SCRIPT_FILENAME'] = $_SERVER['DOCUMENT_ROOT']."/index.php";

    //get $_SERVER['REQUEST_URI']; until the first ? or # character
    $_SERVER['PATH_INFO'] = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
    $_SERVER['PHP_SELF'] = "/index.php".$_SERVER['PATH_INFO'];
}
$vercelURL = getenv('VERCEL_URL');

putenv("APP_URL=https://$vercelURL");

// Forward Vercel requests to public index.
require __DIR__ . "/" . "../public/index.php";

```

**Comment by zainudinnoori** [comment, marked as answer]
**There are two options to fix this issue:** 
Go to Project > Settings > General > Node.js Version Change from 20.x to 18.x

Or 

Update the vercel-php version to use node 18: `"runtime": "vercel-php@0.6.1"`,

**Reply by adamnik101** [reply]
Thanks, i had same issue. This worked, thanks!

[a comment by an automated account is not shown]
